# NUMERICAL LINEAR ALGEBRA 


### Question 1:
In many situations, the way the problem is set up caN have significant effect on susceptibility to round error when we apply standards methods. But fortunately, the are often ways to minimize those effects. For instance, suppose we want to fit a line for the following data

$$
\begin{aligned}
&\begin{array}{|c|c|c|c|c|c|c|c|c|c|c|}
\hline \hline 
x_i & 1.02 & 0.95 & 0.87 & 0.77 & 0.67 & 0.56 & 0.44 & 0.30 & 0.16 & 0.01\\ \hline
y_i & 0.39 & 0.32 & 0.27 & 0.22 & 0.18 & 0.15 & 0.13 & 0.12 & 0.13 & 0.15\\
\hline
\end{array}
\end{aligned}
$$

by solving normal equation $X^T X \begin{pmatrix}m \\ b\end{pmatrix} = X^T Y$. Using python, find $1$-norm and $\infty$-norm condition of the coefficients of the system of normal equations. Explain your results. Here $X$ is the matrix for estimating parameters $m$ and $b$ for the linear model $mx +b = y$ 



In [1]:
import numpy as np

In [2]:
x = np.array([[1,1.02], [1,0.95],[1,0.87],[1,0.77],[1,0.67],[1,0.56],[1,0.44],[1,0.30],[1,0.16],[1,0.01]])
y = np.array([0.39,0.32,0.27,0.22,0.18,0.15,0.13,0.12,0.13,0.15])

In [3]:
H = x.T@x
G = np.linalg.inv(x.T@x)
H_norm1 = np.linalg.norm(H,1) # norm one of xTx
G_norm1 = np.linalg.norm(G,1) # norm one of inverse(xTx)

In [4]:
cond_xTx = H_norm1*G_norm1 # condition number using norm one
cond_xTx

23.440822111977315

In [5]:
V_norm_inf = np.linalg.norm(H,np.inf) # norm infinity of xTx
U_norm1_inf = np.linalg.norm(G,np.inf) # norm infinity of inverse(xTx)

In [6]:
cond_xTx_inf = V_norm_inf*U_norm1_inf # condition number using norm infinity
cond_xTx_inf

23.44082211197732

#### The condition number of the system (using both norm-1 and norm-infinity) is small therefore the system is better-conditioned. 

In [28]:
m_c = x.T@y
m_c

array([2.06  , 1.4262])

### m = 2.06, c = 1.4262

## Question 2

Solve the following systems of equations using numpy


$$
\begin{eqnarray}
A= \begin{cases} 
2x_1 + x_2 + x_3 &= 4\\
x_1 + 3x_2 + 2x_3 &= 5\\
x_1 &= 6
\end{cases}\qquad \text{and} \qquad
B= \begin{cases} 
4x + 3y + 2z &= 25\\
-2x + 2y + 3z &= -10\\
3x -5y + 2z &= 6
\end{cases}
\end{eqnarray}
$$


In [8]:
import numpy as np

A = np.array([[2,1,1],[1,3,2],[1,0,0]])
b = np.array([4,5,6]).T

Ain = np.linalg.inv(A)
X = Ain@b

print(' x1 =',X[0], '\n', 'x2 = ', X[1], '\n','x3 = ', X[2])


B = np.array([[4,3,2],[-2,2,3],[3,-5,2]])
a = np.array([25,-10,6])

Bin = np.linalg.inv(B)
X = Bin@b
print('System 2')
print(' x =',X[0], '\n', 'y = ', X[1], '\n','z = ', X[2])

 x1 = 6.0 
 x2 =  15.000000000000007 
 x3 =  -23.000000000000007
System 2
 x = 0.21138211382113814 
 y =  -0.27642276422764234 
 z =  1.9918699186991873


## Question 3

1. Write a python funtion <b>fun_prod</b>  that takes two matrices $A$ and $B$ and return a product $C = A \times B$. If the product is not possible, the function should return "<b><i>Multiplication not possible</i></b>".

 <b>Do not use existing function such as $np.multiply$ or $A@B$</b>
 

In [5]:
import numpy as np
C = np.zeros((A.shape[1],B.shape[1]))
def fun_prod(A,B):
    if A.shape[1] != B.shape[0]:
        print ("dimensionality error")
    else:
        for i in range(len(A)):     
            for k in range(len(A)):
                for j in range(B.shape[1]):
                    C[i,j]+=A[i,k]*B[k,j]
        return C

2. Use the function <b>fun_prod</b> to compute $C = A \times B $ for the following matrices

$$
\begin{eqnarray}
    A = \begin{pmatrix}
    0 & 6 & -2\\
    4 & -3 & 3
    \end{pmatrix},\qquad
    B = \begin{pmatrix}
    4 & -4\\
    1 & 5
    \end{pmatrix}  
\end{eqnarray} 
$$
And

$$
\begin{eqnarray}
    A = \begin{pmatrix}
    3 & 0 & 0\\
    2 & 1 & 0\\
    1 & 4 & 3
    \end{pmatrix},\qquad 
    B = \begin{pmatrix}
    3 & 5 & 1\\
    0 & 1 & 3\\
    0 & 0 & 7
    \end{pmatrix},   
\end{eqnarray}
$$

In [6]:
A = np.array([[0,6,-2],[4,-3,3]])
B = np.array([[4,-4],[1,5]])
fun_prod(A,B)

dimensionality error


In [7]:
A = np.array([[3,0,0],[2,1,0],[1,4,3]])
B = np.array([[3,5,1],[0,1,3],[0,0,7]])
fun_prod(A,B)

array([[ 9., 15.,  3.],
       [ 6., 11.,  5.],
       [ 3.,  9., 34.]])